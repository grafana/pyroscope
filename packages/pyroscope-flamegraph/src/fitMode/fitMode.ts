export const TailMode = 'TAIL';
export const HeadMode = 'HEAD';

export type FitModes = typeof TailMode | typeof HeadMode;

const margin = 3;

/**
 * Returns a text and margin left used to write text into a canvas rectangle
 *
 * @param {FitModes} mode -
 * @param {number} charSize - Size in pixels of an individual character. Assumes it's a monospace font.
 * @param {number} rectWidth - Width in pixels of the rectangle
 * @param {string} fullText - The text that will be first tried.
 * @param {string} shortText - The text that willbe used when fullText can't fit. It's normally a substring of the original text.
 */

interface fitToCanvasRectProps {
  mode: FitModes;

  /** charSize - Size in pixels of an individual character. Assumes it's a monospace font. */
  charSize: number;

  /** Width in pixels of the rectangle */
  rectWidth: number;

  /** The text that will be first tried to fit */
  fullText: string;

  /** The text that willbe used when fullText can't fit. It's normally a substring of the original text. */
  shortText: string;
}

export function fitToCanvasRect({
  mode,
  charSize,
  rectWidth,
  fullText,
  shortText,
}: fitToCanvasRectProps) {
  switch (mode) {
    case TailMode:
      // Case 1:
      // content fits rectangle width
      // | rectangle |
      // | text |
      if (charSize * fullText.length <= rectWidth) {
        // assume it's a monospaced font
        return {
          mode,
          text: fullText,
          marginLeft: margin,
        };
      }

      // assume it's a monospaced font
      // if not the case, use
      // ctx.measureText(shortName).width
      const shortTextWidth = charSize * shortText.length;

      // Case 2:
      // short text fits rectangle width
      // | rectangle |
      // | long_text_text |
      // | shorttext |
      if (shortTextWidth <= rectWidth) {
        // assume it's a monospaced font
        return {
          mode,
          text: shortText,
          marginLeft: margin,
        };
      }

      // Case 3:
      // short text is bigger than rectangle width
      // add a negative margin left
      // so that the 'tail' of the string is visible
      //     | rectangle |
      // | my_short_text |
      return {
        mode,
        text: shortText,
        marginLeft: -(shortTextWidth - rectWidth + margin),
      };

    // Case 3:
    // Normal
    case HeadMode:
    default:
      return {
        mode,
        text: fullText,
        marginLeft: margin,
      };
  }
}

/**
 * Returns an inline style in React format
 * used to fit the content into a table cell
 * or an empty object if not applicable.
 * @param {FitModes} mode - The mode
 */
export function fitIntoTableCell(mode: FitModes): React.CSSProperties {
  switch (mode) {
    case TailMode:
      return {
        // prints from right to left
        direction: 'rtl',
        overflow: 'hidden',
        textOverflow: 'ellipsis',
      };

    case HeadMode:
    default:
      return {
        overflow: 'hidden',
        textOverflow: 'ellipsis',
      };
  }
}
