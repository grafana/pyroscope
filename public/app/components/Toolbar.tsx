import React from 'react';
import 'react-dom';

import { useAppSelector, useAppDispatch } from '@pyroscope/redux/hooks';
import { Query } from '@pyroscope/models/query';
import {
  selectApps,
  reloadAppNames,
  selectQueries,
  selectAppNamesState,
} from '@pyroscope/redux/reducers/continuous';
import { faSyncAlt } from '@fortawesome/free-solid-svg-icons/faSyncAlt';
import Button from '@pyroscope/ui/Button';
import LoadingSpinner from '@pyroscope/ui/LoadingSpinner';
import DateRangePicker from '@pyroscope/components/DateRangePicker';
import RefreshButton from '@pyroscope/components/RefreshButton';
import { AppSelector } from '@pyroscope/components/AppSelector/AppSelector';
import styles from './Toolbar.module.css';

interface ToolbarProps {
  /** callback to be called when an app is selected via the dropdown */
  onSelectedApp: (name: Query) => void;

  filterApp?: (names: string) => boolean;
}
function Toolbar({
  onSelectedApp,
  filterApp: _filterApp = () => true,
}: ToolbarProps) {
  const dispatch = useAppDispatch();
  const appNamesState = useAppSelector(selectAppNamesState);
  const apps = useAppSelector(selectApps);
  const { query } = useAppSelector(selectQueries);
  const selectedQuery = query;

  const onSelected = (query: Query) => {
    onSelectedApp(query);
  };

  const appNamesLoading =
    appNamesState.type === 'reloading' ? (
      <LoadingSpinner className={styles.appNamesLoading} />
    ) : null;

  return (
    <>
      <div className="navbar">
        <div className={styles.leftSide}>
          <AppSelector
            onSelected={onSelected}
            apps={apps}
            selectedQuery={selectedQuery}
          />
          <Button
            aria-label="Refresh Apps"
            icon={faSyncAlt}
            onClick={() => dispatch(reloadAppNames())}
            className={styles.refreshAppsButton}
          />
          {appNamesLoading}
        </div>
        <div className="navbar-space-filler" />
        <RefreshButton />
        &nbsp;
        <DateRangePicker />
      </div>
    </>
  );
}

export default Toolbar;
