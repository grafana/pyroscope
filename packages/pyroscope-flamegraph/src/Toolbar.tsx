import React, {
  ReactNode,
  RefObject,
  useState,
  useRef,
  useLayoutEffect,
  isValidElement,
  memo,
} from 'react';
import classNames from 'classnames/bind';
import { faUndo } from '@fortawesome/free-solid-svg-icons/faUndo';
import { faCompressAlt } from '@fortawesome/free-solid-svg-icons/faCompressAlt';
import { faProjectDiagram } from '@fortawesome/free-solid-svg-icons/faProjectDiagram';
import { faEllipsisV } from '@fortawesome/free-solid-svg-icons/faEllipsisV';
import { Maybe } from 'true-myth';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import useResizeObserver from '@react-hook/resize-observer';
// until ui is moved to its own package this should do it
// eslint-disable-next-line import/no-extraneous-dependencies
import Button from '@webapp/ui/Button';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Tooltip } from '@pyroscope/webapp/javascript/ui/Tooltip';
import { FitModes } from './fitMode/fitMode';
import SharedQueryInput from './SharedQueryInput';
import type { ViewTypes } from './FlameGraph/FlameGraphComponent/viewTypes';
import type { FlamegraphRendererProps } from './FlameGraph/FlameGraphRenderer';
import {
  TableIcon,
  TablePlusFlamegraphIcon,
  FlamegraphIcon,
  SandwichIcon,
  HeadFirstIcon,
  TailFirstIcon,
} from './Icons';

import styles from './Toolbar.module.scss';

const cx = classNames.bind(styles);

const DIVIDER_WIDTH = 5;
const QUERY_INPUT_WIDTH = 175;
const LEFT_MARGIN = 2;
const RIGHT_MARGIN = 2;
const TOOLBAR_SQUARE_WIDTH = 40 + LEFT_MARGIN + RIGHT_MARGIN;
const MORE_BUTTON_WIDTH = 16;

const calculateCollapsedItems = (
  clientWidth: number,
  collapsedItemsNumber: number,
  itemsW: number[]
) => {
  const availableToolbarItemsWidth =
    collapsedItemsNumber === 0
      ? clientWidth - QUERY_INPUT_WIDTH - 5
      : clientWidth - QUERY_INPUT_WIDTH - MORE_BUTTON_WIDTH - 5;

  let collapsedItems = 0;
  let visibleItemsWidth = 0;
  itemsW.reverse().forEach((v) => {
    visibleItemsWidth += v;
    if (availableToolbarItemsWidth <= visibleItemsWidth) {
      collapsedItems += 1;
    }
  });

  return collapsedItems;
};

const useMoreButton = (
  target: RefObject<HTMLDivElement>,
  toolbarItemsWidth: number[]
) => {
  const [isCollapsed, setCollapsedStatus] = useState(true);
  const [collapsedItemsNumber, setCollapsedItemsNumber] = useState(0);

  useLayoutEffect(() => {
    if (target.current) {
      const { width } = target.current.getBoundingClientRect();
      const collapsedItems = calculateCollapsedItems(
        width,
        collapsedItemsNumber,
        toolbarItemsWidth
      );
      setCollapsedItemsNumber(collapsedItems);
    }
  }, [target.current, toolbarItemsWidth]);

  const handleMoreClick = () => {
    setCollapsedStatus((v) => !v);
  };

  useResizeObserver(target, (entry: ResizeObserverEntry) => {
    const { width } = entry.target.getBoundingClientRect();
    const collapsedItems = calculateCollapsedItems(
      width,
      collapsedItemsNumber,
      toolbarItemsWidth
    );

    setCollapsedItemsNumber(collapsedItems);
    setCollapsedStatus(true);
  });

  return {
    isCollapsed,
    handleMoreClick,
    collapsedItemsNumber,
  };
};

export interface ProfileHeaderProps {
  view: ViewTypes;
  enableChangingDisplay?: boolean;
  flamegraphType: 'single' | 'double';
  handleSearchChange: (s: string) => void;
  highlightQuery: string;
  ExportData?: ReactNode;

  /** Whether the flamegraph is different from its original state */
  isFlamegraphDirty: boolean;
  reset: () => void;

  updateFitMode: (f: FitModes) => void;
  fitMode: FitModes;
  updateView: (s: ViewTypes) => void;

  /**
   * Refers to the node that has been selected in the flamegraph
   */
  selectedNode: Maybe<{ i: number; j: number }>;
  onFocusOnSubtree: (i: number, j: number) => void;
  sharedQuery?: FlamegraphRendererProps['sharedQuery'];
}

const Divider = () => <div className={styles.divider} />;

type ToolbarItemType = {
  width: number;
  el: ReactNode;
};

const Toolbar = memo(
  ({
    view,
    handleSearchChange,
    highlightQuery,
    isFlamegraphDirty,
    reset,
    updateFitMode,
    fitMode,
    updateView,
    selectedNode,
    onFocusOnSubtree,
    flamegraphType,
    enableChangingDisplay = true,
    sharedQuery,
    ExportData,
  }: ProfileHeaderProps) => {
    const toolbarRef = useRef<HTMLDivElement>(null);

    const fitModeItem = {
      el: (
        <>
          <FitMode fitMode={fitMode} updateFitMode={updateFitMode} />
          <Divider />
        </>
      ),
      width: TOOLBAR_SQUARE_WIDTH * 2 + DIVIDER_WIDTH,
    };
    const resetItem = {
      el: <ResetView isFlamegraphDirty={isFlamegraphDirty} reset={reset} />,
      width: TOOLBAR_SQUARE_WIDTH,
    };
    const focusOnSubtree = {
      el: (
        <>
          <FocusOnSubtree
            selectedNode={selectedNode}
            onFocusOnSubtree={onFocusOnSubtree}
          />
          <Divider />
        </>
      ),
      width: TOOLBAR_SQUARE_WIDTH + DIVIDER_WIDTH,
    };

    const viewSectionItem = enableChangingDisplay
      ? {
          el: (
            <ViewSection
              flamegraphType={flamegraphType}
              view={view}
              updateView={updateView}
            />
          ),
          // sandwich view is hidden in diff view
          width: TOOLBAR_SQUARE_WIDTH * (flamegraphType === 'single' ? 5 : 3), // 1px is to display divider
        }
      : null;
    const exportDataItem = isValidElement(ExportData)
      ? {
          el: (
            <>
              <Divider />
              {ExportData}
            </>
          ),
          width: TOOLBAR_SQUARE_WIDTH + DIVIDER_WIDTH,
        }
      : null;

    const filteredToolbarItems = [
      fitModeItem,
      resetItem,
      focusOnSubtree,
      viewSectionItem,
      exportDataItem,
    ].filter((v) => v !== null) as ToolbarItemType[];
    const toolbarItemsWidth = filteredToolbarItems.reduce(
      (acc, v) => [...acc, v.width],
      [] as number[]
    );

    const { isCollapsed, collapsedItemsNumber, handleMoreClick } =
      useMoreButton(toolbarRef, toolbarItemsWidth);

    const toolbarFilteredItems = filteredToolbarItems.reduce(
      (acc, v, i) => {
        const isHiddenItem = i < collapsedItemsNumber;

        if (isHiddenItem) {
          acc.hidden.push(v);
        } else {
          acc.visible.push(v);
        }

        return acc;
      },
      { visible: [] as ToolbarItemType[], hidden: [] as ToolbarItemType[] }
    );

    return (
      <div role="toolbar" ref={toolbarRef}>
        <div className={styles.navbar}>
          <div>
            <SharedQueryInput
              width={QUERY_INPUT_WIDTH}
              onHighlightChange={handleSearchChange}
              highlightQuery={highlightQuery}
              sharedQuery={sharedQuery}
            />
          </div>
          <div>
            <div className={styles.itemsContainer}>
              {toolbarFilteredItems.visible.map((v, i) => (
                // eslint-disable-next-line react/no-array-index-key
                <div key={i} className={styles.item} style={{ width: v.width }}>
                  {v.el}
                </div>
              ))}
              {collapsedItemsNumber !== 0 && (
                <Tooltip placement="top" title="More">
                  <button
                    onClick={handleMoreClick}
                    className={cx({
                      [styles.moreButton]: true,
                      [styles.active]: !isCollapsed,
                    })}
                  >
                    <FontAwesomeIcon icon={faEllipsisV} />
                  </button>
                </Tooltip>
              )}
            </div>
          </div>
          {!isCollapsed && (
            <div className={styles.navbarCollapsedItems}>
              {toolbarFilteredItems.hidden.map((v, i) => (
                <div
                  // eslint-disable-next-line react/no-array-index-key
                  key={i}
                  className={styles.item}
                  style={{ width: v.width }}
                >
                  {v.el}
                </div>
              ))}
            </div>
          )}
        </div>
      </div>
    );
  }
);

function FocusOnSubtree({
  onFocusOnSubtree,
  selectedNode,
}: {
  selectedNode: ProfileHeaderProps['selectedNode'];
  onFocusOnSubtree: ProfileHeaderProps['onFocusOnSubtree'];
}) {
  const onClick = selectedNode.mapOr(
    () => {},
    (f) => {
      return () => onFocusOnSubtree(f.i, f.j);
    }
  );

  return (
    <Tooltip placement="top" title="Collapse nodes above">
      <div>
        <Button
          disabled={!selectedNode.isJust}
          onClick={onClick}
          className={styles.collapseNodeButton}
          aria-label="Collapse nodes above"
        >
          <FontAwesomeIcon icon={faCompressAlt} />
        </Button>
      </div>
    </Tooltip>
  );
}

function ResetView({
  isFlamegraphDirty,
  reset,
}: {
  isFlamegraphDirty: ProfileHeaderProps['isFlamegraphDirty'];
  reset: ProfileHeaderProps['reset'];
}) {
  return (
    <Tooltip placement="top" title="Reset View">
      <span>
        <Button
          id="reset"
          disabled={!isFlamegraphDirty}
          onClick={reset}
          className={styles.resetViewButton}
          aria-label="Reset View"
        >
          <FontAwesomeIcon icon={faUndo} />
        </Button>
      </span>
    </Tooltip>
  );
}

function FitMode({
  fitMode,
  updateFitMode,
}: {
  fitMode: ProfileHeaderProps['fitMode'];
  updateFitMode: ProfileHeaderProps['updateFitMode'];
}) {
  const isSelected = (a: FitModes) => fitMode === a;

  return (
    <>
      <Tooltip placement="top" title="Head first">
        <Button
          onClick={() => updateFitMode('HEAD')}
          className={cx({
            [styles.fitModeButton]: true,
            [styles.selected]: isSelected('HEAD'),
          })}
        >
          <HeadFirstIcon />
        </Button>
      </Tooltip>
      <Tooltip placement="top" title="Tail first">
        <Button
          onClick={() => updateFitMode('TAIL')}
          className={cx({
            [styles.fitModeButton]: true,
            [styles.selected]: isSelected('TAIL'),
          })}
        >
          <TailFirstIcon />
        </Button>
      </Tooltip>
    </>
  );
}

const getViewOptions = (
  flamegraphType: ProfileHeaderProps['flamegraphType']
): Array<{
  label: string;
  value: ViewTypes;
  Icon: (props: { fill?: string | undefined }) => JSX.Element;
}> =>
  flamegraphType === 'single'
    ? [
        { label: 'Table', value: 'table', Icon: TableIcon },
        {
          label: 'Table and Flamegraph',
          value: 'both',
          Icon: TablePlusFlamegraphIcon,
        },
        {
          label: 'Flamegraph',
          value: 'flamegraph',
          Icon: FlamegraphIcon,
        },
        { label: 'Sandwich', value: 'sandwich', Icon: SandwichIcon },
        {
          label: 'GraphViz',
          value: 'graphviz',
          Icon: () => <FontAwesomeIcon icon={faProjectDiagram} />,
        },
      ]
    : [
        { label: 'Table', value: 'table', Icon: TableIcon },
        {
          label: 'Table and Flamegraph',
          value: 'both',
          Icon: TablePlusFlamegraphIcon,
        },
        {
          label: 'Flamegraph',
          value: 'flamegraph',
          Icon: FlamegraphIcon,
        },
      ];

function ViewSection({
  view,
  updateView,
  flamegraphType,
}: {
  updateView: ProfileHeaderProps['updateView'];
  view: ProfileHeaderProps['view'];
  flamegraphType: ProfileHeaderProps['flamegraphType'];
}) {
  const options = getViewOptions(flamegraphType);

  return (
    <div className={styles.viewType}>
      {options.map(({ label, value, Icon }) => (
        <Tooltip key={value} placement="top" title={label}>
          <Button
            data-testid={value}
            onClick={() => updateView(value)}
            className={cx({
              [styles.toggleViewButton]: true,
              selected: view === value,
            })}
          >
            <Icon />
          </Button>
        </Tooltip>
      ))}
    </div>
  );
}

export default Toolbar;
