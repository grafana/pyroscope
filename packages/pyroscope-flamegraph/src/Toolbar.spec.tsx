import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Maybe } from 'true-myth';
import Toolbar from './Toolbar';
import { HeadMode, TailMode } from './fitMode/fitMode';

// since 'react-debounce-input' uses lodash.debounce under the hood
jest.mock('lodash.debounce', () =>
  jest.fn((fn) => {
    fn.flush = () => {};
    return fn;
  })
);

describe('ProfileHeader', () => {
  beforeAll(() => {
    window.HTMLElement.prototype.getBoundingClientRect = function () {
      return {
        x: 0,
        y: 0,
        bottom: 0,
        right: 0,
        toJSON: () => {},
        height: 0,
        top: 0,
        left: 0,
        width: 900,
      };
    };
  });

  it('should render toolbar correctly', () => {
    const { asFragment } = render(
      <Toolbar
        view="both"
        flamegraphType="single"
        handleSearchChange={() => {}}
        reset={() => {}}
        updateFitMode={() => {}}
        fitMode={HeadMode}
        updateView={() => {}}
        isFlamegraphDirty={false}
        selectedNode={Maybe.nothing()}
        onFocusOnSubtree={() => {}}
        highlightQuery=""
      />
    );

    expect(screen.getByRole('toolbar')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Reset button', () => {
    const onReset = jest.fn();

    beforeEach(() => {});

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('renders as disabled when flamegraph is not dirty', () => {
      const component = (
        <Toolbar
          view="both"
          flamegraphType="single"
          isFlamegraphDirty={false}
          handleSearchChange={() => {}}
          reset={onReset}
          updateFitMode={() => {}}
          fitMode={HeadMode}
          updateView={() => {}}
          selectedNode={Maybe.nothing()}
          onFocusOnSubtree={() => {}}
          highlightQuery=""
        />
      );
      render(component);
      expect(screen.getByRole('button', { name: /Reset/ })).toBeDisabled();
    });

    it('calls onReset when clicked (and enabled)', () => {
      const component = (
        <Toolbar
          view="both"
          flamegraphType="single"
          isFlamegraphDirty
          handleSearchChange={() => {}}
          reset={onReset}
          updateFitMode={() => {}}
          fitMode={HeadMode}
          updateView={() => {}}
          selectedNode={Maybe.nothing()}
          onFocusOnSubtree={() => {}}
          highlightQuery=""
        />
      );
      render(component);
      expect(screen.getByRole('button', { name: /Reset/ })).not.toBeDisabled();
      screen.getByRole('button', { name: /Reset/ }).click();

      expect(onReset).toHaveBeenCalled();
    });
  });

  describe('HighlightSearch', () => {
    it('calls callback when typed', () => {
      const onChange = jest.fn();

      const component = (
        <Toolbar
          view="both"
          flamegraphType="single"
          isFlamegraphDirty
          handleSearchChange={onChange}
          reset={() => {}}
          updateFitMode={() => {}}
          fitMode={HeadMode}
          updateView={() => {}}
          selectedNode={Maybe.nothing()}
          onFocusOnSubtree={() => {}}
          highlightQuery=""
        />
      );

      render(component);
      userEvent.type(screen.getByRole('searchbox'), 'foobar');
      expect(onChange).toHaveBeenCalledWith('foobar');
    });
  });

  describe('FitMode', () => {
    const updateFitMode = jest.fn();
    const component = (
      <Toolbar
        view="both"
        flamegraphType="single"
        handleSearchChange={() => {}}
        reset={() => {}}
        updateFitMode={updateFitMode}
        fitMode={HeadMode}
        updateView={() => {}}
        isFlamegraphDirty={false}
        selectedNode={Maybe.nothing()}
        onFocusOnSubtree={() => {}}
        highlightQuery=""
      />
    );

    beforeEach(() => {
      render(component);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('updates to HEAD first', () => {
      screen.getByRole('button', { name: 'Head first' }).click();

      expect(updateFitMode).toHaveBeenCalledWith(HeadMode);
    });

    it('updates to TAIL first', () => {
      screen.getByRole('button', { name: 'Tail first' }).click();

      expect(updateFitMode).toHaveBeenCalledWith(TailMode);
    });
  });

  describe('Focus on subtree', () => {
    it('renders as disabled when theres no selected node', () => {
      const component = (
        <Toolbar
          view="both"
          flamegraphType="single"
          isFlamegraphDirty={false}
          handleSearchChange={() => {}}
          reset={() => {}}
          updateFitMode={() => {}}
          fitMode={HeadMode}
          updateView={() => {}}
          selectedNode={Maybe.nothing()}
          onFocusOnSubtree={() => {}}
          highlightQuery=""
        />
      );
      render(component);
      expect(screen.getByRole('button', { name: /Collapse/ })).toBeDisabled();
    });

    it('calls callback when clicked', () => {
      const onFocusOnSubtree = jest.fn();
      const component = (
        <Toolbar
          view="both"
          flamegraphType="single"
          isFlamegraphDirty={false}
          handleSearchChange={() => {}}
          reset={() => {}}
          updateFitMode={() => {}}
          fitMode={HeadMode}
          updateView={() => {}}
          selectedNode={Maybe.just({ i: 999, j: 999 })}
          onFocusOnSubtree={onFocusOnSubtree}
          highlightQuery=""
        />
      );

      render(component);
      screen.getByRole('button', { name: /Collapse/ }).click();

      expect(onFocusOnSubtree).toHaveBeenCalledWith(999, 999);
    });
  });
});
