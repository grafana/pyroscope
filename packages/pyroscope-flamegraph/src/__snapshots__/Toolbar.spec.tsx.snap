// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileHeader shifts between visualization modes 1`] = `
<DocumentFragment>
  <div
    data-mode="large"
    role="toolbar"
  >
    <div
      class="navbar"
    >
      <input
        class="search "
        data-testid="flamegraph-search"
        name="flamegraph-search"
        placeholder="Search…"
        type="search"
        value=""
      />
      <div
        class="space-filler"
      />
      <select
        aria-label="fit-mode"
        class="fit-mode-select"
      >
        <option
          disabled=""
        >
          Prefer to Fit
        </option>
        <option
          value="HEAD"
        >
          Head first
        </option>
        <option
          value="TAIL"
        >
          Tail first
        </option>
      </select>
      <button
        class="button  default undefined"
        data-testid="reset-view"
        disabled=""
        id="reset"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-undo fa-w-16 iconWithText"
          data-icon="undo"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
            fill="currentColor"
          />
        </svg>
        Reset View
      </button>
      <button
        class="button  default undefined"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-compress-alt fa-w-14 iconWithText"
          data-icon="compress-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"
            fill="currentColor"
          />
        </svg>
        Focus on subtree
      </button>
      <div
        class="btn-group viz-switch"
      >
        <button
          class="button grouped default undefined"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-table fa-w-16 iconWithText"
            data-icon="table"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64v-96h160v96zm0-160H64v-96h160v96zm224 160H288v-96h160v96zm0-160H288v-96h160v96z"
              fill="currentColor"
            />
          </svg>
          Table
        </button>
        <button
          class="button grouped primary undefined"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-columns fa-w-16 iconWithText"
            data-icon="columns"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM224 416H64V160h160v256zm224 0H288V160h160v256z"
              fill="currentColor"
            />
          </svg>
          Both
        </button>
        <button
          class="button grouped default undefined"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-icicles fa-w-16 iconWithText"
            data-icon="icicles"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M511.4 37.9C515.1 18.2 500 0 480 0H32C10.6 0-4.8 20.7 1.4 41.2l87.1 273.4c2.5 7.2 12.7 7.2 15.1 0L140 190.5l44.2 187.3c1.9 8.3 13.7 8.3 15.6 0l46.5-196.9 34.1 133.4c2.3 7.6 13 7.6 15.3 0l45.8-172.5 66.7 363.8c1.7 8.6 14 8.6 15.7 0l87.5-467.7z"
              fill="currentColor"
            />
          </svg>
          Flamegraph
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ProfileHeader shifts between visualization modes 2`] = `
<DocumentFragment>
  <div
    data-mode="small"
    role="toolbar"
  >
    <div
      class="navbar"
    >
      <input
        class="search search-small"
        data-testid="flamegraph-search"
        name="flamegraph-search"
        placeholder="Search…"
        type="search"
        value=""
      />
      <div
        class="space-filler"
      />
      <select
        aria-label="fit-mode"
        class="fit-mode-select"
      >
        <option
          disabled=""
        >
          Fit
        </option>
        <option
          value="HEAD"
        >
          Head
        </option>
        <option
          value="TAIL"
        >
          Tail
        </option>
      </select>
      <button
        class="button  default undefined"
        data-testid="reset-view"
        disabled=""
        id="reset"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-undo fa-w-16 iconWithText"
          data-icon="undo"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z"
            fill="currentColor"
          />
        </svg>
        Reset
      </button>
      <button
        class="button  default undefined"
        disabled=""
        type="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-compress-alt fa-w-14 iconWithText"
          data-icon="compress-alt"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"
            fill="currentColor"
          />
        </svg>
        Focus
      </button>
      <div
        class="btn-group viz-switch"
      >
        <select
          aria-label="view"
        >
          <option
            value="table"
          >
            Table
          </option>
          <option
            value="both"
          >
            Both
          </option>
          <option
            value="icicle"
          >
            Flame
          </option>
        </select>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
